src/platform_specific/affinity.unix.cc

```diff
@@ -1,20 +1,226 @@
+	thread_remover& operator=(thread_remover&&) = delete;
+	~thread_remover() {
+		std::lock_guard lock(g_state.mutex);
+		if(g_state.initialized) { g_state.pinned_threads.erase(pthread_self()); }
+	}
+};
+thread_local std::optional<thread_remover> t_remover; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+
+// Initializes the thread pinning machinery
+// This captures the current thread's affinity mask and sets the thread pinning machinery up
+// Calls to pin_this_thread prior to this call will have no effect
+bool initialize(const runtime_configuration& cfg) {
+	std::lock_guard lock(g_state.mutex);
+	if(g_state.initialized) {
+		CELERITY_ERROR("Thread pinning already initialized. Ignoring this initialization attempt.");
+		return false;
+	}
+	assert(g_state.thread_pinning_plan.empty() && "Thread pinning plan not initially empty.");
+	assert(g_state.pinned_threads.empty() && "Pinned threads not initially empty.");
+
+	g_state.config = cfg;
+
+	const auto ret = sched_getaffinity(0, sizeof(cpu_set_t), &g_state.available_cores);
+	if(ret != 0) {
+		CELERITY_WARN("Error retrieving initial process affinity mask. Unable to check whether enough logical cores are available to this process.{}",
+		    cfg.enabled ? " Will disable thread pinning." : "");
+		g_state.config.enabled = false;
+		return true;
+	}
+
+	// pinned threads per process: application, scheduler, executor, 1 device submitter per device if enabled
+	uint32_t pinned_threads_per_process = 3;
+	if(g_state.config.use_backend_device_submission_threads) { pinned_threads_per_process += g_state.config.num_devices; }
+	// total number of threads to be pinned across processes (legacy mode) - we assume that each process has been assigned the same number of device
+	const uint32_t total_threads = pinned_threads_per_process * g_state.config.num_legacy_processes;
+
+	if(g_state.config.enabled) {
+		// select the core set to use
+		std::vector<uint32_t> selected_core_ids = {};
+		if(!cfg.hardcoded_core_ids.empty()) {
+			// attempt to use the provided hardcoded IDs if they match the number of threads to be pinned
+			if(static_cast<uint32_t>(cfg.hardcoded_core_ids.size()) != total_threads) {
+				CELERITY_WARN("Hardcoded core ID count ({}) does not match the number of threads to be pinned ({}), downgrading to auto-pinning.",
+				    cfg.hardcoded_core_ids.size(), total_threads);
+			} else {
+				selected_core_ids = cfg.hardcoded_core_ids;
+			}
+		}
+		if(selected_core_ids.empty()) {
+			// otherwise, sequential core assignments for now; it is most important that each of the threads is "close"
+			// to the ones next to it in this sequence, so that communication between them is fast
+			selected_core_ids = get_available_sequential_cores(g_state.available_cores, total_threads, cfg.standard_core_start_id);
+			if(selected_core_ids.empty()) {
+				CELERITY_WARN("Insufficient available cores for thread pinning (required {}, {} available), disabling pinning.", //
```

Comment:
```suggestion
				CELERITY_WARN("Insufficient logical cores available for thread pinning (required {}, {} available), disabling pinning. Performance may be negatively impacted.", //
```

To bring it a bit more in line with the other warning.

Reply:
Actually the warning can get nonsensical when using the `from:` syntax. I just got this:

![image](https://github.com/user-attachments/assets/37094181-1196-4bf3-b136-1c0519a77d05)

Reply:
I made it a bit more meaningful.